File: frontend/src/types/contentBlocks.ts
export const CONTENT_BLOCK_TYPE = {
    TEXT: 'text',
    VISUAL_WIDGET: 'visual_widget',
} as const;

export type ContentBlockType =
    (typeof CONTENT_BLOCK_TYPE)[keyof typeof CONTENT_BLOCK_TYPE];

export const WIDGET_BLOCK_STATUS = {
    IN_PROGRESS: 'in_progress',
    COMPLETE: 'complete',
    ERROR: 'error',
} as const;

export type WidgetBlockStatus =
    (typeof WIDGET_BLOCK_STATUS)[keyof typeof WIDGET_BLOCK_STATUS];

export interface ContentBlock {
    type: ContentBlockType;
    order: number;
    status: WidgetBlockStatus;
    text?: string | null;
    textChunks?: string[];
    title?: string | null;
    loadingMessages?: string[];
    widgetCode?: string | null;
    widgetCodeChunks?: string[];
    errorMessage?: string | null;
}

export interface BlockOpenPayload {
    event_type: 'block-open';
    type: ContentBlockType;
    order: number;
    status: WidgetBlockStatus;
    content?: string | null;
    error_message?: string | null;
}

export interface BlockDeltaPayload {
    event_type: 'block-delta';
    order: number;
    type: ContentBlockType;
    status: WidgetBlockStatus;
    content?: string | null;
    error_message?: string | null;
}

export interface BlockClosePayload {
    event_type: 'block-close';
    order: number;
    type: ContentBlockType;
    status: WidgetBlockStatus;
    content?: string | null;
    error_message?: string | null;
}

/** Raw block from REST (camelCase after apiFetch) or SSE (snake_case). */
export function apiBlockToContentBlock(raw: {
    type: string;
    order: number;
    status: string;
    text?: string | null;
    module?: string | null;
    title?: string | null;
    loading_messages?: string[];
    loadingMessages?: string[];
    widget_code?: string | null;
    widgetCode?: string | null;
    error_message?: string | null;
    errorMessage?: string | null;
}): ContentBlock {
    return {
        type: raw.type as ContentBlockType,
        order: raw.order,
        status: raw.status as WidgetBlockStatus,
        text: raw.text,
        title: raw.title,
        loadingMessages: raw.loadingMessages ?? raw.loading_messages ?? [],
        widgetCode: raw.widgetCode ?? raw.widget_code,
        errorMessage: raw.errorMessage ?? raw.error_message,
    };
}

File: agent/models/content_blocks.py
from dataclasses import asdict, dataclass
from enum import StrEnum
import json

from sse_starlette import ServerSentEvent
from pydantic import BaseModel


class ContentBlockType(StrEnum):
    TEXT = "text"
    VISUAL_WIDGET = "visual_widget"


class WidgetBlockStatus(StrEnum):
    IN_PROGRESS = "in_progress"
    COMPLETE = "complete"
    ERROR = "error"


class ContentBlockEventType(StrEnum):
    OPEN = "block-open"
    DELTA = "block-delta"
    CLOSE = "block-close"


@dataclass
class ContentBlock:
    id: str
    event_type: ContentBlockEventType
    type: ContentBlockType
    order: int
    status: WidgetBlockStatus
    content: str | None = None
    error_message: str | None = None

    def to_sse(self) -> ServerSentEvent:
        return ServerSentEvent(
            event=self.event_type.value,
            data=json.dumps(asdict(self)),
        )


class PersistedContentBlock(BaseModel):
    id: str
    type: ContentBlockType
    order: int
    status: WidgetBlockStatus
    text: str | None = None
    widget_code: str | None = None
    error_message: str | None = None

    @classmethod
    def from_events(
        cls,
        events: list[ContentBlock],
    ) -> list["PersistedContentBlock"]:
        blocks: dict[int, PersistedContentBlock] = {}
        chunks: dict[int, list[str]] = {}
        for event in events:
            if event.event_type == ContentBlockEventType.OPEN:
                blocks[event.order] = cls(
                    id=event.id,
                    type=event.type,
                    order=event.order,
                    status=event.status,
                    error_message=event.error_message,
                )
                chunks[event.order] = []
                continue
            if event.event_type == ContentBlockEventType.DELTA:
                chunks.setdefault(event.order, []).append(event.content or "")
                continue
            block = blocks.get(event.order)
            if block is not None:
                blocks[event.order] = block.model_copy(
                    update={
                        "status": event.status,
                        "error_message": event.error_message,
                    },
                )

        result: list[PersistedContentBlock] = []
        for order, block in sorted(blocks.items()):
            content = "".join(chunks[order])
            update = (
                {"text": content}
                if block.type == ContentBlockType.TEXT
                else {"widget_code": content}
            )
            result.append(block.model_copy(update=update))
        return result
